rmw_zenoh_cpp/src/detail/rmw_context_impl_s.cpp

```diff
@@ -136,60 +151,58 @@ class rmw_context_impl_s::Data final
     // the code will be simpler, and if we're just going to spin over the non-blocking
     // reads until we obtain responses, we'll just be hogging CPU time by convincing
     // the OS that we're doing actual work when it could instead park the thread.
-    z_owned_reply_channel_t channel = zc_reply_fifo_new(0);
+    z_owned_fifo_handler_reply_t handler;
+    z_owned_closure_reply_t closure;
+    z_fifo_channel_reply_new(&closure, &handler, SIZE_MAX - 1);
+
+
+    z_view_keyexpr_t keyexpr;
+    z_view_keyexpr_from_str(&keyexpr, liveliness_str.c_str());
     zc_liveliness_get(
-      z_loan(session_), z_keyexpr(liveliness_str.c_str()),
-      z_move(channel.send), NULL);
-    z_owned_reply_t reply = z_reply_null();
-    for (bool call_success = z_call(channel.recv, &reply); !call_success || z_check(reply);
-      call_success = z_call(channel.recv, &reply))
-    {
-      if (!call_success) {
-        continue;
-      }
-      if (z_reply_is_ok(&reply)) {
-        z_sample_t sample = z_reply_ok(&reply);
-        z_owned_str_t keystr = z_keyexpr_to_string(sample.keyexpr);
+      z_loan(session_), z_loan(keyexpr),
+      z_move(closure), NULL);
+
+    z_owned_reply_t reply;
+    while (z_recv(z_loan(handler), &reply) == Z_OK) {
+      if (z_reply_is_ok(z_loan(reply))) {
+        const z_loaned_sample_t * sample = z_reply_ok(z_loan(reply));
+        z_view_string_t keystr;
+        z_keyexpr_as_view_string(z_sample_keyexpr(sample), &keystr);
+        std::string str(z_string_data(z_loan(keystr)), z_string_len(z_loan(keystr)));
```

Comment:
`str` is a bit generic.

```suggestion
        std::string liveliness_str(z_string_data(z_loan(keystr)), z_string_len(z_loan(keystr)));
```